### PR TITLE
rewrite with egoscale v3

### PR DESCRIPTION
**Pull request by sauterp**
## Description
We rewrite `snap-o-matic` with egoscale v3.
We disable the CI. It fails due to missing credentials which we don't intend to provide.

## Backwards incompatible changes
- logging to a file or syslog is no longer supported (the flags -l --log have been removed)
- Previously the program checked whether a snapshot was created by `snap-o-matic` before deleting a snapshot. Now it just sorts the snapshots by `created-at` and deletes the oldest ones.

## Testing
I tested running it locally and on an Exoscale compute instance(to make sure that getting the instance ID from the metadata server works.)
```shell
# on a compute instance:
> export EXOSCALE_API_KEY=...
> export EXOSCALE_API_SECRET=...
> export EXOSCALE_API_ENDPOINT=https://api-ch-dk-2.exoscale.com/v2

> ./snap-o-matic
2024/04/29 10:38:55 INFO looking up instance ID via metadata service
2024/04/29 10:38:55 INFO defaulting to host instance instance-id={UUID:228a4dc0-e2c9-480c-a37f-71370360af6d}
2024/04/29 10:38:55 INFO settings config="{APIEndpoint:https://api-ch-dk-2.exoscale.com/v2 SnapshotsRetention:7 DryRun:false InstanceID:{UUID:228a4dc0-e2c9-480c-a37f-71370360af6d} CredentialsFile: LogLevel:info}"
2024/04/29 10:38:55 INFO found snapshot id=8e80d3b6-d5c4-40e9-bf67-ddbd509bfba8 created-at=2024-04-29T10:32:03.000Z
2024/04/29 10:38:55 INFO found snapshot id=7fc55314-ca87-45a8-ad6a-c863a916a77b created-at=2024-04-29T10:30:06.000Z
2024/04/29 10:38:55 INFO found snapshot id=413164aa-13c0-4b07-aa00-c74f14e83263 created-at=2024-04-29T10:07:15.000Z
2024/04/29 10:38:55 INFO found snapshot id=a2d57b5f-ff67-46e0-b01f-9f7b87d264ed created-at=2024-04-29T09:59:24.000Z
2024/04/29 10:38:55 INFO found snapshot id=056270fe-3a95-4536-8aea-d9a8e9c1dbd2 created-at=2024-04-29T09:54:21.000Z
2024/04/29 10:38:55 INFO creating snapshot
2024/04/29 10:38:55 INFO waiting for operation to succeed operation-id=d625f0a3-739e-43a0-b8d5-d027843e773d command=get-snapshot
2024/04/29 10:39:28 INFO snapshot created successfully snapshot-id=d639e428-5b88-4757-a918-693e492984ee
```

`make build` and `docker build .` both work.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -5,61 +5,110 @@ import (
...
+		var err error
+		creds, err = apiCredentialsFromFile(cfg.CredentialsFile)
+		if err != nil {
+			exitWithErr(err)
+		}
+	} else {
+		creds = credentials.NewEnvCredentials()
+	}
+
+	client, err := v3.NewClient(creds, v3.ClientOptWithEndpoint(cfg.APIEndpoint))
```
Comment (pierre-emmanuelJ) [inline main.go:69]: A new client is already by default in gva2, just for your info :)

File: main.go [hunk 2]
```diff
@@ -5,61 +5,110 @@ import (
...
 	metadataURL = "http://169.254.169.254/latest/meta-data/vm-id"
-	autosnapTag = "autosnap"
 
 	defaultSnapshotsRetention = 7
-)
-
-var (
-	exo *egoscale.Client
-	cfg config
+	defaultEndpoint           = v3.CHGva2
```
Comment (pierre-emmanuelJ) [inline main.go:25]: By default, v3 returns a client in gva2 (according to API spec), but it sounds you are using it for the usage.

File: main.go [hunk 3]
```diff
@@ -5,61 +5,110 @@ import (
...
-	"github.com/exoscale/egoscale"
 	flag "github.com/spf13/pflag"
-	log "gopkg.in/inconshreveable/log15.v2"
+
+	v3 "github.com/exoscale/egoscale/v3"
+	"github.com/exoscale/egoscale/v3/credentials"
 )
 
 const (
 	metadataURL = "http://169.254.169.254/latest/meta-data/vm-id"
```
Comment (pierre-emmanuelJ) [inline main.go]:
```suggestion
	metadataURL = "http://metadata.exoscale.com/latest/meta-data/vm-id"
```

Comment (pierre-emmanuelJ) [reply main.go]: https://community.exoscale.com/documentation/compute/cloud-init/#querying-the-user-data-and-meta-data-from-the-instance

File: main.go [hunk 4]
```diff
@@ -84,209 +133,161 @@ API credentials file format:
...
+`, defaultEndpoint)
 	}
 
 	flag.Parse()
+}
 
-	logLevelHandler, err := log.LvlFromString(logLevel)
+// rotateSnapshots lists the existing instance snapshots and deletes the oldest ones in order to remain under the
+// specified retention threshold.
+// WARNING: unlike previous versions of snap-o-matic, this will not preserve user-created snapshots
```
Comment (pierre-emmanuelJ) [inline main.go:144]: Thanks for the note

**Review by pierre-emmanuelJ — APPROVED** [review]
Thanks for the work @sauterp 
I put a hold on the PR until I've not done testing on my side then :)

File: main.go (hunk 2 above)
Comment (sauterp) [reply main.go:25]: good to know, but yes in this case it wouldn't make the code simpler if I relied on the v3 default.

File: main.go (hunk 4 above)
Comment (sauterp) [reply main.go:144]: I added the note to the README aswell.

**Comment by pierre-emmanuelJ** [comment]
For me, you can go, just tried to run it multiple time from a newly created instance:
```Shell
ubuntu@test:~/snap-o-matic$ ./bin/snap-o-matic
2024/04/29 15:11:57 INFO looking up instance ID via metadata service
2024/04/29 15:11:57 INFO defaulting to host instance instance-id={UUID:2efdbd5c-5ba3-4481-abfc-080fcb8acfc5}
2024/04/29 15:11:57 INFO settings config="{APIEndpoint:https://api-ch-gva-2.exoscale.com/v2 SnapshotsRetention:7 DryRun:false InstanceID:{UUID:2efdbd5c-5ba3-4481-abfc-080fcb8acfc5} CredentialsFile: LogLevel:info}"
2024/04/29 15:11:57 INFO found snapshot id=55edaee4-5d7a-4bfc-b198-20f9056e0355 created-at=2024-04-29T15:10:17.000Z
2024/04/29 15:11:57 INFO creating snapshot
2024/04/29 15:11:57 INFO waiting for operation to succeed operation-id=b65988b5-a90c-48b2-a54c-9e469fb0da7c command=get-snapshot
2024/04/29 15:12:27 INFO snapshot created successfully snapshot-id=0aefbe5e-5eae-44b7-a853-f3abd00cc959
ubuntu@test:~/snap-o-matic$
```

You can ship for me :shipit:

**Review by pierre-emmanuelJ — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -7,6 +7,16 @@ This is experimental software and may not work as intended or may not be continu
 Use at your own risk. 
 ```
 
+```
+*** WARNING ***
+
+Unlike previous versions of snap-o-matic, v0.02 and later will not preserve user-created snapshots.
```
Comment (arnaudgeiser) [inline README.md:13]: :+1:

File: main.go [hunk 6]
```diff
@@ -84,209 +133,161 @@ API credentials file format:
...
-		dieOnError("invalid value for option --log-level: %s", err)
+		return fmt.Errorf("failed to list snapshots: %w", err)
 	}
 
-	logHandler = egoScaleLog.GetLogHandler(logTo)
-
-	log.Root().SetHandler(log.LvlFilterHandler(logLevelHandler, logHandler))
+	// sort in ascending order, oldest snapshot first
+	slices.SortFunc(snapshots.Snapshots, func(a, b v3.Snapshot) int {
+		return -a.CreatedAT.Compare(b.CreatedAT)
```
Comment (arnaudgeiser) [inline main.go:153]: :+1:

**Review by arnaudgeiser — APPROVED** [review]

**Pull request merged**